examples/siamese_network/siamese_network.py

```diff
@@ -0,0 +1,322 @@
+
+        return anchor_image, positive_image, negative_image, anchor_label
+
+
+def pairwise_distance(input1, input2):
+    dist = input1 - input2
+    dist = torch.pow(dist, 2)
+    return dist
+
+
+def calculate_loss(input1, input2):
+    output = pairwise_distance(input1, input2)
+    loss = torch.sum(output, 1)
+    loss = torch.sqrt(loss)
+    return loss
+
+
+def make_predictions(loss, anchor_label, other_label):
+    pred = torch.where(loss < 3, 1, 0)
+    correct = 0
+    for i in range(loss.shape[0]):
+        if anchor_label[i].item() == other_label[i].item():
+            if pred[i] == 1:
+                correct += 1
+        if anchor_label[i].item() != other_label[i].item():
+            if pred[i] == 0:
+                correct += 1
+    return correct
+
+
+def run(args, model, device, optimizer, train_loader, test_loader, lr_scheduler):
+
+    # using Triplet Margin Loss
+    criterion = nn.TripletMarginLoss(p=2, margin=2.8)
+    total_correct = 0
+
+    # define model training step
+    def train_step(engine, batch):
+        model.train()
+        anchor_image, positive_image, negative_image, anchor_label = batch
+        anchor_image = anchor_image.to(device)
+        positive_image, negative_image = positive_image.to(device), negative_image.to(device)
+        anchor_label = anchor_label.to(device)
+        optimizer.zero_grad()
+        anchor_out, positive_out, negative_out = model(anchor_image, positive_image, negative_image)
+        loss = criterion(anchor_out, positive_out, negative_out)
+        loss.backward()
+        optimizer.step()
+        return loss
+
+    # define model testing step
+    def test_step(engine, batch):
+        model.eval()
+        anchor_image, _, _, anchor_label = batch
+        anchor_image = anchor_image.to(device)
+        anchor_label = anchor_label.to(device)
+        other_image = []
+        other_label = []
+        for i in range(anchor_image.shape[0]):
+            index = torch.randint(0, anchor_image.shape[0], (1,)).item()
+            img = anchor_image[index]
+            label = anchor_label[index]
+            other_image.append(img)
+            other_label.append(label)
+        other = torch.stack(other_image)
+        other_label = torch.tensor(other_label)
+        other, other_label = other.to(device), other_label.to(device)
+        anchor_out, other_out, _ = model(anchor_image, other, other)
+        test_loss = calculate_loss(anchor_out, other_out)
+        correct = make_predictions(test_loss, anchor_label, other_label)
+        return correct
+
+    # create engines for trainer and evaluator
+    trainer = Engine(train_step)
+    evaluator = Engine(test_step)
+    evaluator.run(test_loader)
+
+    # attach Running Average Loss metric to trainer and evaluator engines
+    RunningAverage(output_transform=lambda x: x).attach(trainer, "loss")
+    RunningAverage(output_transform=lambda x: x).attach(evaluator, "loss")
```

Comment:
Does it make sense to compute running average loss for evaluator ?

Reply:
Sorry I realize it now that the `RunningAverage` is not a feasible metric since the function was returning the number of `correct` predictions. I have changed the metric to accuracy for the `evaluator` engine.